File: treessence/src/main/java/fr/bipi/tressence/file/FileLoggerTree.java
package fr.bipi.tressence.file;

import android.util.Log;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.logging.FileHandler;
import java.util.logging.Formatter;
import java.util.logging.Level;
import java.util.logging.LogRecord;
import java.util.logging.Logger;

import fr.bipi.tressence.base.FormatterPriorityTree;
import fr.bipi.tressence.base.NoTree;
import fr.bipi.tressence.common.formatter.LogcatFormatter;
import fr.bipi.tressence.common.utils.FileUtils;
import timber.log.Timber;

/**
 * An implementation of `Timber.Tree` which sends log into a circular file.
 * <p>It is using {@link java.util.logging.Logger} to implement circular file logging
 */
public class FileLoggerTree extends FormatterPriorityTree {
    private static final String TAG = "FileLoggerTree";
    private final Logger logger;
    private final FileHandler fileHandler;
    private final String path;
    private final int nbFiles;

    /**
     * @param priority    Priority from which to log
     * @param logger      {@link java.util.logging.Logger} used for logging
     * @param fileHandler {@link java.util.logging.FileHandler} used for logging
     * @param path        Base path of file
     * @param nbFiles     Max number of files
     */
    private FileLoggerTree(int priority,
                           Logger logger,
                           FileHandler fileHandler,
                           String path, int nbFiles) {
        super(priority);
        this.logger = logger;
        this.fileHandler = fileHandler;
        this.path = path;
        this.nbFiles = nbFiles;
    }

    @Override
    protected void log(int priority, String tag, @NotNull String message, Throwable t) {
        if (skipLog(priority, tag, message, t)) {
            return;
        }

        logger.log(fromPriorityToLevel(priority), format(priority, tag, message));
        if (t != null) {
            logger.log(fromPriorityToLevel(priority), "", t);
        }
    }

    @Override
    protected fr.bipi.tressence.common.formatter.Formatter getDefaultFormatter() {
        return LogcatFormatter.INSTANCE;
    }

    /**
     * Delete all log files
     */
    public void clear() {
        if (fileHandler != null) {
            fileHandler.close();
        }
        for (int i = 0; i < nbFiles; i++) {
            File f = new File(getFileName(i));
            if (f.exists() && f.isFile()) {
                //noinspection ResultOfMethodCallIgnored
                f.delete();
            }
        }
    }

    /**
     * Return the file name corresponding to the number
     *
     * @param i Number of file
     * @return Real file name
     */
    public String getFileName(int i) {
        String path;
        if (!this.path.contains("%g")) {
            path = this.path + "." + i;
        } else {
            path = this.path.replace("%g", "" + i);
        }
        return path;
    }

    /**
     * @return All files created by the logger
     */
    public Collection<File> getFiles() {
        Collection<File> col = new ArrayList<>(nbFiles);
        for (int i = 0; i < nbFiles; i++) {
            File f = new File(getFileName(i));
            if (f.exists()) {
                col.add(f);
            }
        }
        return col;
    }

    private Level fromPriorityToLevel(int priority) {
        switch (priority) {
            case Log.VERBOSE:
                return Level.FINER;
            case Log.DEBUG:
                return Level.FINE;
            case Log.INFO:
                return Level.INFO;
            case Log.WARN:
                return Level.WARNING;
            case Log.ERROR:
                return Level.SEVERE;
            case Log.ASSERT:
                return Level.SEVERE;
            default:
                return Level.FINEST;
        }
    }

    @SuppressWarnings("WeakerAccess")
    public static class Builder {
        private static final int SIZE_LIMIT = 1048576;
        private static final int NB_FILE_LIMIT = 3;

        private String fileName = "log";
        private String dir = "";
        private int priority = Log.INFO;
        private int sizeLimit = SIZE_LIMIT;
        private int fileLimit = NB_FILE_LIMIT;
        private boolean appendToFile = true;

        /**
         * Specify a custom file name
         * <p> Default file name is "log" which will result in log.0, log.1, log.2...
         *
         * @param fn File name
         * @return itself
         */
        public Builder withFileName(String fn) {
            this.fileName = fn;
            return this;
        }

        /**
         * Specify a custom dir name
         *
         * @param dn Dir name
         * @return itself
         */
        public Builder withDirName(String dn) {
            this.dir = dn;
            return this;
        }

        /**
         * Specify a custom dir name
         *
         * @param d Dir file
         * @return itself
         */
        public Builder withDir(File d) {
            this.dir = d.getAbsolutePath();
            return this;
        }

        /**
         * Specify a priority from which it can start logging
         * <p> Default is {@link Log#INFO}
         *
         * @param p priority
         * @return itself
         */
        public Builder withMinPriority(int p) {
            this.priority = p;
            return this;
        }

        /**
         * Specify a custom file size limit
         * <p> Default is 1048576 bytes
         *
         * @param nbBytes Custom size limit in bytes
         * @return itself
         */
        public Builder withSizeLimit(int nbBytes) {
            this.sizeLimit = nbBytes;
            return this;
        }

        /**
         * Specify a custom file number limit
         * <p> Default is 3
         *
         * @param f Max number of files to use
         * @return itself
         */
        public Builder withFileLimit(int f) {
            this.fileLimit = f;
            return this;
        }

        /**
         * Specify an option for {@link FileHandler} creation
         *
         * @param b true to append to existing file
         * @return itself
         */
        public Builder appendToFile(boolean b) {
            this.appendToFile = b;
            return this;
        }

        /**
         * Create the file logger tree with options specified.
         *
         * @return {@link FileLoggerTree} or {@link NoTree} if an exception occurred
         */
        @NotNull
        public Timber.Tree build() {
            Timber.Tree tree;
            String path = FileUtils.combinePath(dir, fileName);

            try {
                FileHandler fileHandler;
                Logger logger = MyLogger.getLogger(TAG);
                if (logger.getHandlers().length == 0) {
                    fileHandler = new FileHandler(path, sizeLimit, fileLimit, appendToFile);
                    fileHandler.setFormatter(new NoFormatter());
                    logger.addHandler(fileHandler);
                } else {
                    fileHandler = (FileHandler) logger.getHandlers()[0];
                }
                tree = new FileLoggerTree(priority, logger, fileHandler, path, fileLimit);
            } catch (IOException e) {
                Timber.e(e);
                tree = new NoTree();
            }

            return tree;
        }
    }

    private static class NoFormatter extends Formatter {

        @Override
        public String format(LogRecord record) {
            return record.getMessage();
        }
    }

    /**
     * Custom logger class that has no references to LogManager
     */
    private static class MyLogger extends Logger {

        /**
         * Constructs a {@code Logger} object with the supplied name and resource
         * bundle name; {@code notifyParentHandlers} is set to {@code true}.
         * <p/>
         * Notice : Loggers use a naming hierarchy. Thus "z.x.y" is a child of "z.x".
         *
         * @param name the name of this logger, may be {@code null} for anonymous
         *             loggers.
         */
        MyLogger(String name) {
            super(name, null);
        }

        public static Logger getLogger(String name) {
            return new MyLogger(name);
        }
    }
}

File: treessence/src/main/java/fr/bipi/tressence/file/FileLoggerTree.kt
package fr.bipi.tressence.file

import android.util.Log
import fr.bipi.tressence.base.FormatterPriorityTree
import fr.bipi.tressence.base.NoTree
import fr.bipi.tressence.common.filters.Filter
import fr.bipi.tressence.common.filters.NoFilter
import fr.bipi.tressence.common.formatter.Formatter
import fr.bipi.tressence.common.formatter.LogcatFormatter
import fr.bipi.tressence.common.utils.FileUtils
import timber.log.Timber
import java.io.File
import java.io.IOException
import java.util.*
import java.util.logging.FileHandler
import java.util.logging.Level
import java.util.logging.LogRecord
import java.util.logging.Logger


/**
 * An implementation of `Timber.Tree` which sends log into a circular file.
 *
 * It is using [java.util.logging.Logger] to implement circular file logging
 *
 * @param priority    Priority from which to log
 * @param logger      [java.util.logging.Logger] used for logging
 * @param fileHandler [java.util.logging.FileHandler] used for logging
 * @param path        Base path of file
 * @param nbFiles     Max number of files
 */
@Suppress("unused")
class FileLoggerTree @JvmOverloads constructor(
    private val logger: Logger,
    private val fileHandler: FileHandler?,
    private val path: String,
    private val nbFiles: Int,
    priority: Int,
    filter: Filter = NoFilter.INSTANCE,
    formatter: Formatter = LogcatFormatter.INSTANCE
) : FormatterPriorityTree(priority, filter, formatter) {

    override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
        if (skipLog(priority, tag, message, t)) {
            return
        }
        logger.log(fromPriorityToLevel(priority), format(priority, tag, message))
        if (t != null) {
            logger.log(fromPriorityToLevel(priority), "", t)
        }
    }

    /**
     * Delete all log files
     */
    fun clear() {
        fileHandler?.close()
        for (i in 0 until nbFiles) {
            val f = File(getFileName(i))
            if (f.exists() && f.isFile) {
                f.delete()
            }
        }
    }

    /**
     * Return the file name corresponding to the number
     *
     * @param i Number of file
     * @return Real file name
     */
    fun getFileName(i: Int): String {
        return if (!this.path.contains("%g")) {
            this.path + "." + i
        } else {
            this.path.replace("%g", "" + i)
        }
    }

    /**
     * @return All files created by the logger
     */
    val files: Collection<File>
        get() {
            val col: MutableCollection<File> = ArrayList(nbFiles)
            for (i in 0 until nbFiles) {
                val f = File(getFileName(i))
                if (f.exists()) {
                    col.add(f)
                }
            }
            return col
        }

    private fun fromPriorityToLevel(priority: Int): Level {
        return when (priority) {
            Log.VERBOSE -> Level.FINER
            Log.DEBUG -> Level.FINE
            Log.INFO -> Level.INFO
            Log.WARN -> Level.WARNING
            Log.ERROR -> Level.SEVERE
            Log.ASSERT -> Level.SEVERE
            else -> Level.FINEST
        }
    }

    class Builder {
        private var fileName = "log"
        private var dir = ""
        private var priority = Log.INFO
        private var sizeLimit = SIZE_LIMIT
        private var fileLimit = NB_FILE_LIMIT
        private var appendToFile = true
        private var filter: Filter = NoFilter.INSTANCE
        private var formatter: Formatter = LogcatFormatter.INSTANCE

        /**
         * Specify a custom file name
         *
         *  Default file name is "log" which will result in log.0, log.1, log.2...
         *
         * @param fn File name
         * @return itself
         */
        fun withFileName(fn: String): Builder {
            fileName = fn
            return this
        }

        /**
         * Specify a custom dir name
         *
         * @param dn Dir name
         * @return itself
         */
        fun withDirName(dn: String): Builder {
            dir = dn
            return this
        }

        /**
         * Specify a custom dir name
         *
         * @param d Dir file
         * @return itself
         */
        fun withDir(d: File): Builder {
            dir = d.absolutePath
            return this
        }

        /**
         * Specify a priority from which it can start logging
         *
         *  Default is [Log.INFO]
         *
         * @param p priority
         * @return itself
         */
        fun withMinPriority(p: Int): Builder {
            priority = p
            return this
        }

        /**
         * Specify a custom file size limit
         *
         *  Default is 1048576 bytes
         *
         * @param nbBytes Custom size limit in bytes
         * @return itself
         */
        fun withSizeLimit(nbBytes: Int): Builder {
            sizeLimit = nbBytes
            return this
        }

        /**
         * Specify a custom file number limit
         *
         *  Default is 3
         *
         * @param f Max number of files to use
         * @return itself
         */
        fun withFileLimit(f: Int): Builder {
            fileLimit = f
            return this
        }

        /**
         * Specify a log filter
         *
         * @param filter Log filter
         */
        fun withFilter(filter: Filter): Builder {
            this.filter = filter
            return this
        }

        /**
         * Specify a log formatter
         *
         * @param formatter Log formatter
         */
        fun withFormatter(formatter: Formatter): Builder {
            this.formatter = formatter
            return this
        }

        /**
         * Specify an option for [FileHandler] creation
         *
         * @param b true to append to existing file
         * @return itself
         */
        fun appendToFile(b: Boolean): Builder {
            appendToFile = b
            return this
        }

        /**
         * Create the file logger tree with options specified.
         *
         * @return [FileLoggerTree]
         * @throws IOException if file creation fails
         */
        @Throws(IOException::class)
        fun build(): FileLoggerTree {
            val path = FileUtils.combinePath(dir, fileName)
            val fileHandler: FileHandler
            val logger = MyLogger.getLogger(TAG)
            logger.level = Level.ALL
            if (logger.handlers.isEmpty()) {
                fileHandler = FileHandler(path, sizeLimit, fileLimit, appendToFile)
                fileHandler.formatter = NoFormatter()
                logger.addHandler(fileHandler)
            } else {
                fileHandler = logger.handlers[0] as FileHandler
            }
            return FileLoggerTree(logger, fileHandler, path, fileLimit, priority, filter, formatter)
        }

        /**
         * Create the file logger tree with options specified.
         *
         * @return [FileLoggerTree] or [NoTree] if an exception occurred
         */
        fun buildQuietly(): Timber.Tree {
            return try {
                build()
            } catch (e: IOException) {
                Timber.e(e)
                NoTree()
            }
        }

        companion object {
            const val SIZE_LIMIT = 1048576
            const val NB_FILE_LIMIT = 3
        }
    }

    private class NoFormatter : java.util.logging.Formatter() {
        override fun format(record: LogRecord): String {
            return record.message
        }
    }

    /**
     * Custom logger class that has no references to LogManager
     *
     * Constructs a `Logger` object with the supplied name and resource
     * bundle name; `notifyParentHandlers` is set to `true`.
     *
     *
     * Notice : Loggers use a naming hierarchy. Thus "z.x.y" is a child of "z.x".
     *
     * @param name the name of this logger, may be `null` for anonymous
     * loggers.
     */
    private class MyLogger(name: String?) : Logger(name, null) {
        companion object {
            fun getLogger(name: String?): Logger {
                return MyLogger(name)
            }
        }
    }

    companion object {
        private const val TAG = "FileLoggerTree"
    }
}
